### PR TITLE
Sw add coinbase checking

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This PR adds in the checking of the block validity before adding in a block the db.

## Motivation and Context
We need to know if blocks are valid by testing the following:
1. Sum up to zero
2. Includes the correct coinbase amount
3. Utxo maturities have been set and are correct

## How Has This Been Tested?
This breaks none of the current unit tests. 
Block validation has just been linked into the base node

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by SWvheerden** [comment]
@neonknight64 We need to go over the unit test values and ensure that they are still correct. 
This PR made all blocks bigger by 1 output and 1 kernel.

**Comment by SWvheerden** [comment]
#941

File: base_layer/core/src/chain_storage/test/chain_storage.rs [hunk 1]
```diff
@@ -620,10 +620,10 @@ fn restore_mmr() {
...
     let header_mmr_state = store.fetch_mmr_base_leaf_nodes(MmrTree::Header, 0, 100).unwrap();
 
-    assert_eq!(utxo_mmr_state.total_leaf_count, 9);
-    assert_eq!(kernel_mmr_state.total_leaf_count, 3);
-    assert_eq!(rp_mmr_state.total_leaf_count, 9);
-    assert_eq!(header_mmr_state.total_leaf_count, 2);
+    assert_eq!(utxo_mmr_state.total_leaf_count, 10);
+    assert_eq!(kernel_mmr_state.total_leaf_count, 4); // Todo This might be wrong, verify  1 + 3 + 2 +2  = 8
+    assert_eq!(rp_mmr_state.total_leaf_count, 10);
+    assert_eq!(header_mmr_state.total_leaf_count, 2); // Todo This might be wrong, verify 4
```
Comment (neonknight64) [inline base_layer/core/src/chain_storage/test/chain_storage.rs]:
```suggestion
    assert_eq!(header_mmr_state.total_leaf_count, 2); 
```
A header leaf count of 2 is correct. It is only the total leaf count of the base mmr. Only the genesis block and block 1 would be part of the base mmr at this point due to the set MerkleChangeTracker settings.

File: base_layer/core/src/chain_storage/test/chain_storage.rs [hunk 2]
```diff
@@ -620,10 +620,10 @@ fn restore_mmr() {
     let rp_mmr_state = store.fetch_mmr_base_leaf_nodes(MmrTree::RangeProof, 0, 100).unwrap();
     let header_mmr_state = store.fetch_mmr_base_leaf_nodes(MmrTree::Header, 0, 100).unwrap();
 
-    assert_eq!(utxo_mmr_state.total_leaf_count, 9);
-    assert_eq!(kernel_mmr_state.total_leaf_count, 3);
-    assert_eq!(rp_mmr_state.total_leaf_count, 9);
-    assert_eq!(header_mmr_state.total_leaf_count, 2);
+    assert_eq!(utxo_mmr_state.total_leaf_count, 10);
+    assert_eq!(kernel_mmr_state.total_leaf_count, 4); // Todo This might be wrong, verify  1 + 3 + 2 +2  = 8
```
Comment (neonknight64) [inline base_layer/core/src/chain_storage/test/chain_storage.rs]:
```suggestion
    assert_eq!(kernel_mmr_state.total_leaf_count, 4); 
```
Only base mmr so 1(GB) + 3(B1)=4

File: base_layer/core/src/chain_storage/test/chain_storage.rs [hunk 3]
```diff
@@ -646,10 +646,10 @@ fn restore_mmr() {
...
         .unwrap()
         .total_leaf_count;
-    assert_eq!(utxo_mmr_leaf_count, 11);
-    assert_eq!(kernel_mmr_leaf_count, 5);
-    assert_eq!(rp_mmr_leaf_count, 11);
-    assert_eq!(header_mmr_leaf_count, 4);
+    assert_eq!(utxo_mmr_leaf_count, 14);
+    assert_eq!(kernel_mmr_leaf_count, 8); // Todo This might be wrong, verify 8 + 2 +2 =12
+    assert_eq!(rp_mmr_leaf_count, 14);
+    assert_eq!(header_mmr_leaf_count, 4); // Todo This might be wrong, verify 6
```
Comment (neonknight64) [inline base_layer/core/src/chain_storage/test/chain_storage.rs]:
```suggestion
    assert_eq!(header_mmr_leaf_count, 4); 
```
Only the first 4 blocks would have been committed to the base mmr at this point.

File: base_layer/core/src/chain_storage/test/chain_storage.rs [hunk 4]
```diff
@@ -646,10 +646,10 @@ fn restore_mmr() {
         .fetch_mmr_base_leaf_nodes(MmrTree::Header, 0, 100)
         .unwrap()
         .total_leaf_count;
-    assert_eq!(utxo_mmr_leaf_count, 11);
-    assert_eq!(kernel_mmr_leaf_count, 5);
-    assert_eq!(rp_mmr_leaf_count, 11);
-    assert_eq!(header_mmr_leaf_count, 4);
+    assert_eq!(utxo_mmr_leaf_count, 14);
+    assert_eq!(kernel_mmr_leaf_count, 8); // Todo This might be wrong, verify 8 + 2 +2 =12
```
Comment (neonknight64) [inline base_layer/core/src/chain_storage/test/chain_storage.rs]:
```suggestion
    assert_eq!(kernel_mmr_leaf_count, 8); 
```
Only 4 block in base mmr. 4(cb of gb,b1,b2 and b3)+4(tx1,tx2,tx3 and tx4)=8

File: base_layer/core/src/chain_storage/test/chain_storage.rs [hunk 5]
```diff
@@ -608,6 +607,7 @@ fn restore_mmr() {
     assert!(store.commit(txn).is_ok());
 
     let mut block1 = chain_block(&block0, vec![tx1.clone(), tx2.clone()]);
+
```
Comment (neonknight64) [inline base_layer/core/src/chain_storage/test/chain_storage.rs]:
```suggestion
```

File: base_layer/core/src/base_node/test/service.rs [hunk 6]
```diff
@@ -522,23 +522,23 @@ fn request_and_response_fetch_mmr_state() {
...
 
         let received_mmr_state = alice_interfaces
             .outbound_nci
             .fetch_mmr_state(MmrTree::Kernel, 1, 2)
             .await
             .unwrap();
-        assert_eq!(received_mmr_state.total_leaf_count, 1);
-        assert_eq!(received_mmr_state.leaf_nodes.leaf_hashes.len(), 0); // request out of range
+        assert_eq!(received_mmr_state.total_leaf_count, 4);
+        assert_eq!(received_mmr_state.leaf_nodes.leaf_hashes.len(), 2); // request out of range
```
Comment (neonknight64) [inline base_layer/core/src/base_node/test/service.rs]:
```suggestion
        assert_eq!(received_mmr_state.leaf_nodes.leaf_hashes.len(), 2); 
```

**Comment by CjS77** [comment]
Will cherry-pick and fold into the validation pipeline code

**Comment by SWvheerden** [comment]
Done in #1095

**Pull request closed without merging**